src/Voron/Impl/Journal/JournalReader.cs

```diff
@@ -128,100 +131,114 @@ public bool ReadOneTransactionToDataFile(StorageEnvironmentOptions options)
 
-            for (var i = 0; i < current->PageCount; i++)
+            long bufferSize = 0;
+            byte* currentBuffer = null;
+            try
             {
-                if (totalRead > current->UncompressedSize)
-                    throw new InvalidDataException($"Attempted to read position {totalRead} from transaction data while the transaction is size {current->UncompressedSize}");
-
-                Debug.Assert(_journalPager.Disposed == false);
-                if (performDecompression)
-                    Debug.Assert(_recoveryPager.Disposed == false);
-
-                var numberOfPagesOnDestination = GetNumberOfPagesFor(pageInfoPtr[i].Size);
-                _dataPager.EnsureContinuous(pageInfoPtr[i].PageNumber, numberOfPagesOnDestination);
-                _dataPager.EnsureMapped(this, pageInfoPtr[i].PageNumber, numberOfPagesOnDestination);
+                for (var i = 0; i < current->PageCount; i++)
+                {
+                    if (totalRead > current->UncompressedSize)
+                        throw new InvalidDataException(
+                            $"Attempted to read position {totalRead} from transaction data while the transaction is size {current->UncompressedSize}");
 
+                    Debug.Assert(_journalPagerState.Disposed == false);
+                    Debug.Assert(performDecompression == false || recoveryPagerState.Disposed == false);
 
-                // We are going to overwrite the page, so we don't care about its current content
-                var pagePtr = _dataPager.AcquirePagePointerForNewPage(this, pageInfoPtr[i].PageNumber, numberOfPagesOnDestination);
-                _dataPager.MaybePrefetchMemory(pageInfoPtr[i].PageNumber, numberOfPagesOnDestination);
-                
-                var pageNumber = *(long*)(outputPage + totalRead);
-                if (pageInfoPtr[i].PageNumber != pageNumber)
-                    throw new InvalidDataException($"Expected a diff for page {pageInfoPtr[i].PageNumber} but got one for {pageNumber}");
-                totalRead += sizeof(long);
+                    var numberOfPagesOnDestination = GetNumberOfPagesFor(pageInfoPtr[i].Size);
+                    _dataPager.EnsureContinuous(ref dataPagerState, pageInfoPtr[i].PageNumber, numberOfPagesOnDestination);
 
-                _modifiedPages.Add(pageNumber);
+                    var pageNumber = *(long*)(outputPage + totalRead);
+                    if (pageInfoPtr[i].PageNumber != pageNumber)
+                        throw new InvalidDataException($"Expected a diff for page {pageInfoPtr[i].PageNumber} but got one for {pageNumber}");
+                    totalRead += sizeof(long);
 
-                for (var j = 1; j < numberOfPagesOnDestination; j++)
-                {
-                    _modifiedPages.Remove(pageNumber + j);
-                }
+                    _modifiedPages.Add(pageNumber);
 
-                _dataPager.UnprotectPageRange(pagePtr, (ulong)pageInfoPtr[i].Size);
```

Comment:
We don't have this validation available any longer? Is it possible to have similar mechanism since we no longer write via mmaps?

Reply:
BTW the project does not compile in VALIDATE mode

Reply:
For most of it, no longer make sense, since we are using read only memory map